clang/test/Analysis/fread.c

```diff
@@ -0,0 +1,412 @@
+    long p = c[0];
+    clang_analyzer_isTainted(p); // expected-warning {{NO}}
+    clang_analyzer_dump(p); // expected-warning {{derived_}}
+
+    p = c[3];
+    clang_analyzer_isTainted(p); // expected-warning {{NO}}
+    clang_analyzer_dump(p); // expected-warning {{derived_}}
+
+    p = c[1];
+    clang_analyzer_isTainted(p); // expected-warning {{YES}}
+    clang_analyzer_dump(p); // expected-warning {{derived_}}
+
+    fclose(fp);
+  }
+}
+
+void dynamic_random_access_write(int startIndex) {
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    long buffer[10];
+    // Cannot reason about index.
+    size_t res = fread(buffer + startIndex, sizeof(long), 5, fp);
+    if (5 == res) {
+      long p = buffer[startIndex];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+    } else if (res == 4) {
+      long p = buffer[startIndex];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[startIndex + 1];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[startIndex + 2];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[startIndex + 3];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[startIndex + 4];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[startIndex + 5];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[0];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+    } else {
+      long p = buffer[startIndex];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+      p = buffer[0];
+      clang_analyzer_isTainted(p); // expected-warning {{NO}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+    }
+    fclose(fp);
+  }
+}
+
+struct S {
+  int a;
+  long b;
+};
+
+void compound_write1(void) {
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    struct S s; // s.a is not touched by fread.
+    if (1 == fread(&s.b, sizeof(s.b), 1, fp)) {
+      long p = s.b;
+      clang_analyzer_isTainted(p); // expected-warning {{YES}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+    } else {
+      long p = s.b;
+      clang_analyzer_isTainted(p); // expected-warning {{YES}}
+      clang_analyzer_dump(p); // expected-warning {{conj_}}
+    }
+    fclose(fp);
+  }
+}
+
+void compound_write2(void) {
```

Comment:
It is not clear to me why some of the test functions are called "write" and some "read". It would be better to have all of them `read` in the name.

Reply:
Probably the author thought that the `read()` function call will have a `write` effect to the buffer.
TBH I agree with you and I'll rename these.